### PR TITLE
Data structure refactor for RPE, RB, and qubit spectroscopy

### PR DESCRIPTION
**Pull request by kylegulshen**
This is my start to the data structure refactor that replaces use of pandas dataframes. Perhaps the most significant change is moving from calling run in the module and receiving bit results to relying on pyquil's measure_observables() and working with ExperimentResults which report operator expectations. Data acquisition methods of separate modules subsequently rely on a shared method in stratified_experiment.py that can accept experiments generated in any of these modules, randomize those experiments, and optionally run some experiments simultaneously. 

There are lots of issues that need smoothing out. Some of the broader structural issues that come to mind:

- ~results are stored by mutating the experiment object.~ The previous philosophy in forest-benchmarking was to copy the whole experiment and return the copy with the additional data. Pyquil's operator_expectation returns a separate results object. This PR currently follows the 'previous philosophy'
- for some modules higher-level estimates and errors are stored in a general-purpose 'estimates' dictionary with a string key describing what it is; I imagine there is a better way? (also similar concerns as above. And do we even want to store estimates in the experiment data structure in the first place?)
- ~independent of this refactor qubit spectroscopy used different methods/modules for plotting data than rb. We want to standardize this--how does that fit in with the above?~ Do we want to store entire models of the fit with stats info? A fitting method essentially serves as the estimate method; we may want to rename these methods to mirror x_estimate.
- how do we want to group experiments to run simultaneously, and how do we report what was done to the user or give the user more control? ~Is the notion of running two Layers simultaneously too coarse?~ Do we need to think about scheduling--we could try to weave in multiple different experiments on the rest of the chip while we have a really long tomography experiment running on a subset of qubits. Is the level of control over randomization sufficient?
- a ramsey experiment with a continuous angle parameter or a t1 with a continuous time parameter doesn't neatly fit into a discrete depth abstraction (I discretized the t1 times). Further, if we use depth as a proxy for experiment run-time for the purposes of grouping things simultaneously (as I've done) then associating depth with angle in a ramsey experiment no longer works because all ramsey experiments essentially take 1 gate-time. Are these differences sufficient to merit a separate data structure? If so, would that sacrifice ease of simultaneously running experiments of different type? 
- a StratifiedExperiment is essentially just a re-packaged list of TomographyExperiments with programs that change slightly with some parameter (depth); should we make that connection more explicit? Is it worth re-packaging Tomography and DFE inside a StratifiedExperiment with a single layer of depth 1? Note that TomographyExperiments cannot be run simultaneously as-is.
- ~experiments like Ramsey/Rabi can naturally be written as parametrized programs but we are missing out on that functionality--is that important? Can it be incorporated easily?~ Maybe for a later release, but this involves fairly major re-writes.

There are probably lots of other things, and I am sure I have made many sub-optimal design decisions so please comment on and directly change anything and everything you see fit from small to large.

### CODE REVIEW

File: forest/benchmarking/qubit_spectroscopy.py [hunk 1]
```diff
@@ -24,165 +31,95 @@
...
+    x_data = [layer.depth * USEC_PER_DEPTH for layer in experiment.layers]  # times in u-seconds
+    y_data = [layer.estimates["Fraction One"][0] for layer in experiment.layers]
+
+    fit_params, fit_params_errs = fit_to_exponential_decay_curve(np.array(x_data), np.array(y_data))
+    #TODO: check if estimates exists?
+    experiment.estimates = {"T1": (fit_params[1], fit_params_errs[1])}
+    return fit_params, fit_params_errs
+
+
+def plot_t1_estimate_over_data(experiments: Union[StratifiedExperiment,
```
Comment (marcusps) [inline forest/benchmarking/qubit_spectroscopy.py]: Shouldn't this be in the plotting submodule?

File: forest/benchmarking/robust_phase_estimation.py [hunk 2]
```diff
@@ -714,64 +434,130 @@ def robust_phase_estimate(experiment: DataFrame, results_label="Results") -> Uni
...
 
 
 #########
 # Plotting
 #########
 
-
-def plot_rpe_iterations(experiments: DataFrame, expected_positions: List = None) -> plt.Axes:
+# TODO: provide more convenient entry for 2q experiments (xs,ys etc. are easy to get for 1q)
+def plot_rpe_iterations(xs, ys, x_stds, y_stds, expected_positions: List = None) -> plt.Axes:
```
Comment (marcusps) [inline forest/benchmarking/robust_phase_estimation.py]: Move to plotting submodule?

**Comment by marcusps** [comment]
@kylegulshen regarding 

> Note that TomographyExperiments cannot be run simultaneously as-is.

I think that is because `TomographyExperiments` is unfinished.  My hypothesis is that the `qubits` field was meant to serve precisely that purpose -- to indicate over which experiment the experiments were meant to run, so that it could be run in parallel with other (non-overlapping) experiments.

**Comment by kylegulshen** [comment]
Sure, that's plausible. Note that merging two programs generically can be a bit dicey though. Consider a narrow Program with a long DELAY paired with a wide shallow process tomography. Also two Experiments could nominally act on the same qubits but still be parallelizable--e.g. drive a particular 2q gate and do RPE on each individual qubit in the chip to estimate the effective RZ. The latter example illustrates that instantiating a TomographyExperiment with a combined program from the outset acting on the union of the qubits and localized ExperimentSettings that can be decomposed into a 'TPB' achieves the same parallelization. 

Regardless, acquire_stratified_data() is meant to deal with these issues. We clearly need the functionality of combining different experiments for benchmarking, and there we may have more information about the type of experiments being run. The code can ultimately live in Pyquil but we need to sort these things out.

File: forest/benchmarking/tests/test_stratified_experiment.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+from pyquil.gates import X, Y
+
+
+def test_merge_sequences():
+    random.seed(0)
+    seq0 = [Program(X(0)), Program(Y(0)), Program(X(0))]
+    seq1 = [Program(X(1)), Program(Y(1)), Program(Y(1))]
+    assert merge_sequences([seq0, seq1]) == [Program(X(0), X(1)),
+                                             Program(Y(0), Y(1)),
+                                             Program(X(0), Y(1))]
```
Comment (blakejohnson) [inline forest/benchmarking/tests/test_stratified_experiment.py]: For such a complicated module, I would expect a much richer set of tests.

**Comment by blakejohnson** [comment]
To the question:
> a StratifiedExperiment is essentially just a re-packaged list of TomographyExperiments with programs that change slightly with some parameter (depth); should we make that connection more explicit? Is it worth re-packaging Tomography and DFE inside a StratifiedExperiment with a single layer of depth 1? Note that TomographyExperiments cannot be run simultaneously as-is.

In general I would not introduction extra wrapping layers without a strong reason to do so. Use the simplest thing possible that will do the job.

**Comment by blakejohnson** [comment]
I'm rather concerned with the complexity of the automatic scheduling of simultaneous experiments within a `StratifiedExperiment`. It also seems to be over used, or trying to solve for too many use cases. e.g. the example @kylegulshen gave in the PR header about T1 as a `StratifiedExperiment`. Is there a document where you all have written down the problem you are trying to solve?

**Comment by joshcombes** [comment]
@blakejohnson I think we should move this discussion an in person discussion

File: examples/direct_fidelity_estimation.ipynb [hunk 4]
```diff
@@ -222,22 +222,9 @@
   }
  ],
  "metadata": {
```
Comment (marcusps) [inline examples/direct_fidelity_estimation.ipynb]: It does not look like there are meaningful changes to this file relative to the PR -- probably we should just remove it.

File: forest/benchmarking/stratified_experiment.py [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+    if parallelize_layers:
+        # by default only layers of the same expt type will possibly be parallelized.
+        expts_by_type_group = _group_allowed_types(experiments, allowed_parallel_types)
+    else:
+        # each layer of each experiment will be run iteratively
+        expts_by_type_group = [[expt] for expt in experiments]
+
+    # for each depth get a list of groups of layers from experiment types that can be parallelized
+    parallelizable_groups_by_depth = _group_by_depth(expts_by_type_group)
+    # TODO: any reason to supply option for not randomizing order of depths?
```
Comment (marcusps) [inline forest/benchmarking/stratified_experiment.py]: We should at least have an option to add a seed for any randomization -- it is nice to have a way to reproduce behavior (if it helps to uncover or demonstrate a bug).

**Comment by kylegulshen** [comment]
Another approach was taken and several PRs are open which replace this one.

**Pull request closed without merging**